mountpoint-s3/src/inode.rs

```diff
@@ -1803,6 +1827,39 @@ mod tests {
         assert_eq!(libc::ENOENT, err, "lookup should return no existing entry error");
     }
 
+    #[tokio::test]
+    async fn test_unlink_verify_checksum() {
+        let client_config = MockClientConfig {
+            bucket: "test_bucket".to_string(),
+            part_size: 1024 * 1024,
+        };
+        let client = Arc::new(MockClient::new(client_config));
+
+        let superblock = Superblock::new("test_bucket", &Default::default(), Default::default());
+
+        let file_name = "corrupted";
+        client.add_object(file_name.as_ref(), MockObject::constant(0xaa, 30, ETag::for_tests()));
+
+        let parent_ino = FUSE_ROOT_INODE;
+        let LookedUp { inode, .. } = superblock
+            .lookup(&client, parent_ino, file_name.as_ref())
+            .await
+            .expect("file should exist");
+
+        // Inject a key mutation in the Inode.
+        // SAFETY: Inode and String not accessed concurrently. Mutated string is still valid utf8.
```

Comment:
This is actually undefined behavior and so not safe (mutating data to which a shared reference exists is [always UB](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)). The risk here is low in practice — the way this would go wrong is if an extremely smart compiler previously cached the `full_key` in a register somewhere, which would be OK because it's a shared immutable field, and so didn't observe this mutation to the key the next time it read it. But in theory the compiler is free to do whatever it wants with the entire program when it sees this.

As an alternative, maybe you could manually construct a new `Inode` struct here as a copy of the existing one, and then overwrite it in the right places (the superblock and the parent of this inode)? It's a lot more coupled to the internals, but maybe unavoidable.